### ISSUE TITLE
Jumps to labels with the same name as a variable / definition

### ISSUE THREAD

**Issue by DanilaFe** [chapel-lang/chapel#21712]
Brad points out the following program in which the behavior of labels can be seen as strange:
```Chapel
label foo for i in 1..10 {
  var foo: int;
  writeln(i);
  if i == 5 then
    break foo;
}
```

The code works (in that it prints the numbers one through five), but should it?
* Do labels get their own namespace, and do they therefore not clash with identifiers in scope (`foo` in this case?)
* If that is the case, should this namespacing behavior be documented somewhere in the spec?
* If labels and variables belong to the same namespace, the code here should error.

**Comment by bradcray** [comment]
Thanks for beating me to filing this Daniel!  (I set a reminder for tomorrow, but might've ended up snoozing it for several days before getting to it).

In most cases in the past where questions have come up like "Should Chapel use a different namespace for modules?" or other symbol types, our answer has been "no, that seems unnecessarily complex / potentially confusing."  My gut reaction to this question is the same—I'd prefer if the behavior were to complain that the labeled break is not referring to a label statement.

**Comment by bradcray** [comment]
Notably, with a quick check in the language spec, I'm not seeing any mention of labels getting a distinct namespace, so this arguably could be considered a bug.  But if it were something we were able to resolve and error for properly in dyno / prior to 2.0, all the better.

**Comment by bradcray** [comment]
@mppf: Do you have an opinion on this one?

**Comment by mppf** [comment]
> * Do labels get their own namespace,

I don't think that they should. However it wouldn't bother me much if they did, as I'd view that as more a curiosity / corner case. If we ever got to the point where you could break to a label computed at runtime, then something like `break` would have to be able to refer to a local variable. But, I don't imagine we need such a feature, so I don't think there would be any serious problem with labels getting their own namespace, other than consistency.

A side issue -- I am uncertain of where `label foo` is located in terms of scoping -- is it in the same scope as `var foo: int` or in a parent scope of it? Either way, this code would result in an error if they are in the same namespace. If the label is considered inside the loop body, it would be a multiple-definition error. If the label is considered outside the loop body, it would be passing an `int` to `break`.

**Comment by bradcray** [comment]
Thanks Michael.  Consistency is my main reason for wanting them to share the same namespace as everything else.  I'm also comfortable with documenting that this is what should happen and calling this a bug (but will leave the 2.0 label on since we don't document it either way currently... though I think the lack of documentation could also be considered an indication that they're not special w.r.t. having their own namespace.

My intuition would be that since the loop being labeled introduces its own scope for the body and the label physically appears to be outside of that scope that it wouldn't be in the same scope as `var foo: int` in the OP (so yeah, I'd expect that the correct behavior would be an error saying that `break` must take a label as its argument, not a variable / an int).  Of course, other patterns could result in multiple-definition errors, like:

```chapel
var foo: int;
label foo for i in ... {
}
```

**Comment by lydia-duncan** [comment]
I think we've decided that this is a bug, is it okay to label it as such and remove the 2.0 label?

**Comment by mppf** [comment]
Calling it a bug sounds good to me.

**Comment by bradcray** [comment]
I renamed the issue to make it less of a question since I think we've taken our stand here.

**Comment by DanilaFe** [comment]
> Bug: Labels for break/continue should get their own namespace, but don't currently

It's late and I might be wrong, but shouldn't it be the other way? I thought we decided that labels should share a namespace with everything else:

* Michael said:
  > I don't think that they should [get their own namespace]. 
* Brad said:
  >  Consistency is my main reason for wanting them to share the same namespace as everything else.
